### DISCUSSION TITLE
Dynamically removing a CartesianLayer Results in an Error

### DISCUSSION THREAD

**Discussion by Tyler-Lopez** [patrykandpatrick/vico#895] [Questions] [answered]
### Question

Hi there!

I am solving for the use-case of wrapping this library in a function, where I provide a series of data. I want to add and remove layers dynamically.

Consider taking the `layers` and making it a remembered list and dynamically adding and removing the second layer in response to a data being removed or added tot he end of the input list.

When I attempt this, I get the following error:

```
java.lang.ClassCastException: com.patrykandpatrick.vico.core.cartesian.layer.LineCartesianLayer cannot be cast to com.patrykandpatrick.vico.core.cartesian.CartesianChart
```

Any thoughts here? Or any thoughts more broadly in how this library could be wrapped to decouple the complicated logic of setting layers and adding lineSeries into just a simple parameter encapsulating both data and style?

### Vico version(s)

2.0.0-alpha.28

### UI framework(s)

Jetpack Compose

**Comment by Gowsky** [comment]
Hello, @Tyler-Lopez.

>I am solving for the use-case of wrapping this library in a function, where I provide a series of data.

This is a specific implementation approach and not a use case. What problem are you trying to solve?

>Consider taking the layers and making it a remembered list and dynamically adding and removing the second layer in response to a data being removed or added tot he end of the input list.

Why is this necessary? The layer won't be drawn if there's no data for it.

>I get the following error:

Please share the entire stack trace and more details in general.

**Reply by Tyler-Lopez** [reply]
Hi there! Thanks so much for the reply, and I apologize for not having the original question clearer.

So, I would say the _use-case_ is that I want to hide and show a set of data. For example, if you are viewing two sets of line data and you want to hide and show the second by an external button. Vico, when used directly, solves this use-case excellently. However, it would seem to require a constant number of `CartesianLayer<*>`.

My desired _implementation_ is to abstract Vico behind a Composable which takes a list of styled data series. 

I'll get you a better example shortly, but it would be something like this:

```kotlin
data class DataSeries(
    val xData: List<Double>,
    val yData: List<Double>,
    val lineColor: Color,
)

@Composable
fun CartesianChart(
    dataSeries: ImmutableList<DataSeries>
) {
    // Remember the model producer
    // Add data series to the model by the model producer
    
    CartesianChartHost(
        chart = rememberCartesianChart(
            layers = dataSeries.map {
                // For each data series, add a line cartesian layer, with the right line color.
            }.toTypedArray()
        ),
        modelProducer = //the remembered model producer,
    )
}
```

In this example, if I update a `DataSeries` object within the parameter `dataSeries` list - it updates excellently. Both the data and the style.

However, if I remove or add a `DataSeries` object, the chart animates as though it was shown for the first time.

This is currently what I'm chewing on. The error in the original message was a result of one attempt to work around this.

I'll think through this a bit more and come back with a clearer non-psuedocoded gist. Thanks again for your thoughts!

**Reply by Gowsky** [reply, marked as answer]
A single `LineCartesianLayer` is enough to display multiple lines with different styles — see sample chart 7. You can change the number of lines at any time by running a `Transaction` that adds or removes series, and you can use `LineProvider` to style the lines dynamically. Adding or removing `CartesianLayer`s isn't necessary.
